text/0001-cdk-update.md

```diff
@@ -0,0 +1,300 @@
+---
+rfc pr: [#335](https://github.com/aws/aws-cdk-rfcs/pull/335)
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/1
+---
+
+# CDK Update
+
+Shorten the development iteration speed -- the "edit-compile-test" loop -- for
+CDK applications.
+
+## Working Backwards
+
+### CHANGELOG
+
+- feat(cli): implement `cdk update` and `cdk update --watch` for asset updates
+
+### README
+
+The `cdk update` command accelerates the edit-compile-test loop for your CDK
+application during development by inspecting the assets and stack resources in
+your application, identifying those that can be updated in-place without a full
+CloudFormation stack udpate, and doing so using AWS service APIs directly. It
+can either do this as an interactive command or by running continuously and
+monitoring the input files.
+
+For supported construct types (see list, below) the update command will identify
+assets whose content has changed since the publication of the CDK, assemble them
+from sources if necessary, and then publish them to their asset storage medium.
+It will then use AWS SDK commands to directly modify the CDK resources to use
+the new asset paths. The update command acts on one or more stacks in the
+application, and will only perform the shortcut if all changes are updateable.
+
+A stack is "updateable" if the stack is explicitly marked as updateable, and it
+contains any assets supported by the update process. At this time, only Lambda
+and ECS images are supported, and so any stack with one of those is updateable.
+
+At a high level, `update` operates by reading the current state of the stack,
+performing a stack diff on it, and if the stack is different only in resources
+supported by update plugins, running those in lieu of a full deployment.
+
+#### Preparing Your Stack
+
+In order to ensure that update is not invoked against production stacks, you
+must explicitly designate your stack as being updateable in your application.
+This is a stack-level property, set in the `StackProps` parameter to the stack
+constructor:
+
+```typescript
+const myDevelopmentStack = new ApplicationStack(app, "StackNameForDev", {
+    ...
+    interactiveUpdate: true,
+});
+```
+
+The `interactiveUpdate` property is optional, and when present and `true` the
+`cdk update` functionality will be executed on the stack.
+
+#### Running Update
+
+To run a one-shot update, you can invoke the `cdk update` command on some or all
+of your stacks:
+
+```
+$ cdk update ApplicationStack
```

Comment:
I don't like the fact that you have to specify the Stack name on the command line, _and_ mark it as `interactiveUpdate: true`. Feels like irritating double-booking to me; why do I have to tell you twice that I want to update this stack with `cdk update`?

I would either have the property on `Stack`, _or_ specifying the names through CLI arguments. Not both.

Reply:
The stack is specified because it's possible to have multiple stacks and only want to update one of them, or to watch one, etc. This, and it matches the usage of stacks elsewhere in the CDK cli. I don't like specifying the stacks either, but it _is_ necessary if you have "stages" of stacks.

Reply:
Then there's no reason for having the `interactiveUpdate` flag at all, and I would remove it.

Reply:
There is, though! Using an at-work example, for a single application that deploys many stacks, only a very small subset of those stacks should be allowed to be updated using this mechanism, and the others need to be denied by policy somewhere. One way that can be done is to explicitly identify which stacks in an app are suitable for it, another is to assume that the user is going to "know" the ones that are permitted in practice. How, otherwise, would a "production" stack prevent a rapid update, if your app consists of multiple stacks, each corresponding to a deployment stage?

Reply:
I'm wondering why we are concerned with which ones are permitted, when we are targeting local developer accounts?